### ISSUE TITLE
Loss cannot drop

### ISSUE THREAD

**Issue by QiushiYang** [lucidrains/vit-pytorch#64]
Thank you so much for sharing your codes. I try to employ Vit as the encoder and follow a common decoder to build a segmentation network. I train it from scratch but found the loss can't drop since the beginning of training, and the results keep near 0. Is there any trick for training Vit correctly? Is it very important to load the pre-train model to fine-tune?
Here is my configuration:
`patch_size=16
hidden_size=16*16*3 
mlp_dim = 3072
dropout_rate = 0.1
num_heads =  12 
num_layers = 12
lr=3e-4
opt=Adam
weight_decay=0.0 `

**Comment by Abhranta** [comment]
I really don't think the mlp return the latent vector that you need to perform a encoder-decoder operation. The mlp head returns the final result for classification tasks.

**Comment by lucidrains** [comment]
@QiushiYang Yup, exactly what @Abhranta said

If you want to modify Vision Transformer for encoder / decoder task, I recommend using https://github.com/lucidrains/x-transformers , which contains the full encoder / decoder, and even comes with a Vision Wrapper so images can be accepted in the encoder

**Comment by lucidrains** [comment]
@QiushiYang see the `Image -> Caption` section, it's exactly what you need

**Comment by Abhranta** [comment]
@lucidrains damn that repo is like a gold mine of information !

**Comment by QiushiYang** [comment]
Thanks a lot for your helps. I will try to remove the conv-based decoder in the segmentation pipeline. Besides, I have another question that can original learnable position embedding (with random initialization) learn the spatial information well in spatial-sensitive tasks such as semantic segmentation?

**Comment by lucidrains** [comment]
@QiushiYang so I think for segmentation you may need better positional encoding than in the ViT paper (although all bets are off at scale).

https://arxiv.org/abs/2005.12872 and https://arxiv.org/abs/2003.07853 are both recent papers that is using attention for object detection / segmentation. The first one actually has an entire section devoted to PE that is worth reading

**Comment by QiushiYang** [comment]
@lucidrains OK. Thanks a lot for your kind suggestions :-)
